src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -3668,13 +3679,55 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
     return filter->process(std::move(bl), lofs);
   }
 
+  void set_in_stream_req(RGWRESTStreamRWRequest *req) override {
+    in_stream_req = req;
+  }
+
+  void make_sse_s3_key(bool *pause) override {
+    if (pause) *pause = true; // xxx: could this race with the spawn? (i.e. pause is set to curl after the spawn)
```

Comment:
could this race with the spawn? (i.e. pause is set to curl after the spawn)

Reply:
looks like it, yeah. `cancel()` is probably okay since it's just using `curl_multi_remove_handle()` and not interacting with the pause stuff. but `unpause_receive()` could potentially race

instead of just passing a `bool *pause` to the callback, maybe we need to pass enough state that the callback can lock and pause itself. then `make_sse_s3_key()` could safely to do that before calling `spawn()` to avoid the race with cancel/unpause

Reply:
wouldn't this by nature again fall into the same deadlock I've experienced (in sending HTTP request to Vault) in my first implementation if the current request was not paused?

Reply:
if this is using the `rgw_http_req_data::mutex`, the vault request would create a different `rgw_http_req_data`

Reply:
sorry, i didn't read `RGWHTTPClient::receive_http_header()` carefully enough. we're already holding that mutex over the call to `req_data->client->receive_header()`, and `req_data->read_paused = true` gets assigned under the same lock

so i'm not sure it _could_ race with the coroutine's call to `RGWHTTPStreamRWRequest::unpause_receive()`. that has to acquire the same mutex, so it would block waiting for `RGWHTTPClient::receive_http_header()` to finish

Reply:
right it's the same `req_data->lock` when we read headers and when we want to unpause. so i guess that could never race. thanks for the hint.